app/src/services/cpk.ts

```diff
@@ -0,0 +1,124 @@
+import { ethers, Wallet } from 'ethers'
+import CPK from 'contract-proxy-kit'
+
+import { getLogger } from '../util/logger'
+import { ConditionalTokenService, ERC20Service, MarketMakerService } from './index'
+import { BigNumber } from 'ethers/utils'
+
+const logger = getLogger('Services::CPKService')
+
+interface CPKBuyOutcomesParams {
+  provider: any
+  cost: BigNumber
+  amount: BigNumber
+  outcomeIndex: number
+  marketMaker: MarketMakerService
+  conditionalTokens: ConditionalTokenService
+}
+
+class CPKService {
+  static buyOutcomes = async ({
+    provider,
+    cost,
+    amount,
+    outcomeIndex,
+    marketMaker,
+    conditionalTokens,
+  }: CPKBuyOutcomesParams) => {
+    try {
+      const signer: Wallet = provider.getSigner()
+      const account = await signer.getAddress()
+
+      const collateralAddress = await marketMaker.getCollateralToken()
+      const marketMakerAddress = marketMaker.address
+      const conditionalTokensAddress = conditionalTokens.address
+
+      const cpk = await CPK.create({ ethers, signer })
+      const cpkAddress = cpk.address
+
+      // Approve amount of collateral to the CPK
+      const collateralService = new ERC20Service(provider, account, collateralAddress)
```

Comment:
I think this shouldn't be done by the CPK service. The callee should do it. Plus, doing it that way will remove a parameter from this function.

Reply:
Mmmm I'm not sure about this, we need the `cpkAddress` to make the approve, and this exist only in the CPK Service, maybe we need a constructor and a property for the address todo this ? what do you think @fvictorio ?

Reply:
We are going to need to have access that address everywhere so we may as well solve this problem now.

Reply:
@fvictorio I move the allowance logic outside the CPK Service